package-type-check/pkg/checkers/biproduct.go

```
@@ -0,0 +1,46 @@
+package checkers
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+// IsPackageInstalled checks if a package is really installed
+func IsSameNamePackageInstalled(pkg string) (bool, error) {
+	cmd := exec.Command("apk", "list", "--installed", pkg)
+	output, err := cmd.Output()
+	if err != nil {
+		return false, fmt.Errorf("failed to get installed version for package %q: %w", pkg, err)
+	}
+
+	// Split the output by lines and get the first line
+	lines := strings.Split(string(output), "\n")
+	if len(lines) == 0 || lines[0] == "" || strings.Compare(lines[0], pkg) != 0 {
+		return false, nil // Package not installed
+	}
+
+	return true, nil
+}
+
+func CheckBiProductPackage(pkg string) error {
+	fmt.Printf("Checking if package %s is a valid bi-product(can't be installed by the package manager) package\n", pkg)
+
+	// Try to install the package
+	cmd := exec.Command("apk", "add", pkg)
+	cmd.Stdout = nil
+	cmd.Stderr = nil
+	_ = cmd.Run()
```

more confusion in my head related to 'BiProductPackage'.
You just ignore the error here, so if it failed then it wont be installed, and it will pass.

Ie, `this-package-does-not-exist` is a valid `BiProductPackage`.

Can you explain what a BiProduct Package is?

---

@smoser 

> Also, as it is used in the pipeline (with `${{context.name}}`), this test will always fail, because the package `${{context.name}}` is guaranteed to be installed by `apko` in the test environment. Right?

**Not always.** If another package with a **higher provider priority** provides the same virtual package, `apko` will resolve and install that one instead.

For example, in the main test pipeline of `py3-pip`, if we inspect the installed packages, we see that `py3.13-pip` is present. This is because it has a higher provider priority than other packages that provide `py3-pip`. So even though the test is running in the context of `py3-pip`, the actual installed package might be different.

---

> You just ignore the error here, so if it failed then it wont be installed, and it will pass.
Ie, this-package-does-not-exist is a valid BiProductPackage.

You're right that a package not being installable can indicate it's a valid by-product package. But I agree with your point about not ignoring the error.

If we blindly ignore installation errors, and no other package with higher provider priority provides this package, then it becomes a genuine failure case that we should catch On the other hand, if a higher-priority provider does not exist, apko will install this package instead, and the test will fail (since the by-product package itself ended up being installed), which is the behavior we will catch during this test.